fxrunner/src/lib/proto.rs

```diff
@@ -106,48 +126,160 @@ where
                 .expect("unzip task was cancelled or panicked");
 
                 if let Err(e) = unzip_result {
-                    self.inner
-                        .send(DownloadBuildReply {
-                            result: Err(e.into_error_message()),
-                        })
-                        .await?;
+                    self.send(DownloadBuildReply {
+                        result: Err(e.into_error_message()),
+                    })
+                    .await?;
 
                     Err(e.into())
                 } else {
                     let firefox_path = download_dir.join("firefox").join("firefox.exe");
 
                     if !firefox_path.exists() {
                         let err = RunnerProtoError::MissingFirefox;
-                        self.inner
-                            .send(DownloadBuildReply {
-                                result: Err(err.into_error_message()),
-                            })
-                            .await?;
+                        self.send(DownloadBuildReply {
+                            result: Err(err.into_error_message()),
+                        })
+                        .await?;
 
                         Err(err)
                     } else {
-                        self.inner
-                            .send(DownloadBuildReply {
-                                result: Ok(DownloadStatus::Extracted),
-                            })
-                            .await?;
+                        self.send(DownloadBuildReply {
+                            result: Ok(DownloadStatus::Extracted),
+                        })
+                        .await?;
 
                         Ok(firefox_path)
                     }
                 }
             }
 
             Err(e) => {
-                error!(self.log, "could not download build"; "error" => %e);
-                self.inner
-                    .send(DownloadBuildReply {
+                error!(self.log, "could not download build"; "error" => ?e);
+                self.send(DownloadBuildReply {
+                    result: Err(e.into_error_message()),
+                })
+                .await?;
+                Err(e.into())
+            }
+        }
+    }
+
+    pub async fn send_profile_reply(
```

Comment:
> 
> 
> So, the code looks good, but at a high level I'm beginning to wonder why we are going with the chatty approach of a predefined sequence of messages sent from the fxrecorder to the fxrunner, rather than a single message with a set of options. I don't see any added flexibility from making it a sequence of messages, and it seems like the protocol very much requires them to be sent in an exact order, so I would think it would be simpler to just send one message, and have the fxrunner report incrementally on its progress?
> 
> What are your thoughts? I just think if we're going to couple all of the messages together, we may as well just make them one message.

I chunked up the API so that the unit tests could be self contained and only test one part each. I'd like to keep it this way at least for all the PRs that are up for review, but I agree its not flexible or elegant.

I'll write up a PR that merges the messages together and redoes the unit tests.

Reply:
> 
> 
> Another retroactive nit / question: does the convention of naming these `<something>_reply` come from somewhere? I personally find it easier to understand the objectives of these functions when they're named something like `recv_<something>`, as replying to the message is not so much the objective as is doing what the message asks. Or maybe something like `expect_<something>` or `handle_<something>`, since we're sort of proactively expecting a particular message sequence?

I guess it might have made more sense to do `<foo>_request` and `<foo>_reply` instead of `<foo>` and `<foo>_reply`. Mostly it was just OK enough and made sense. I'll address this in another PR like I mentioned above.